src/include/firebird/FirebirdInterface.idl

```diff
@@ -1695,3 +1696,49 @@ interface ReplicatedSession : PluginBase
 	void setSequence(Status status, const string name, int64 value);
 }
 
+// Profiler interfaces
+
+interface ProfilerPlugin : PluginBase
+{
+	void init(Status status, Attachment attachment, Transaction transaction);
+
+	ProfilerSession startSession(Status status, Transaction transaction, const string description,
+		ISC_TIMESTAMP_TZ timestamp);
```

Comment:
Why is requires `transaction` ? 
What transaction will be passed into this routine - user tx or separate one ?
How this transaction could be used by plugin ?

Could it be so that plugin creates own metadata using this transaction ? 
Could plugin save own data using this transaction ?

Reply:
It's user transaction and plugin could use it during the call. It should not hold it for later use.

That should be documented.

Reply:
Can I get the answers on two last questions too, pls ?

Reply:
I described things in others answers of what default plugin does with it and here what plugin could not do (hold transaction for later use).

Of course, any prohibited usage thing by Firebird should not be done by plugin - example: create metadata and use in same transaction.

Reply:
Sorry, this is not the answer I could expect.
My point is that good design should nor allow nor provoke users to do wrong things.
So far I see no strong reason to expose user transaction in the profiler plugin interface.

Reply:
The passed transactions could be (and are) used to query if metadata is already created in init and get new sequence in startSession.

Otherwise plugin would need to start transaction to do that. So IMO avoid the transaction does not simplify things nor make things faster.

It's a matter of document how not to use transaction. It's up to plugin to use it in valid situations or not use it.

Reply:
One the one side we have optimization to avoid new transaction when profile session starts.

One the other side we have potential security issue when plugin run own actions in context of user transaction, 
also we have design issue when public API expose too much for external code.

I see no problem having to start and commit new transaction once per profiler session - it is not often and not requires to be highly optimized event.

BTW, I'm not sure profiler metadata should be owned by random user who happens to be first to use profiler. It is better to use separate account for it and make all necessary grants for public (or less wide) usage, IMHO.

Reply:
I agree that metadata creation must not run under random user because in that case we can't correctly control access to it (BTW, that user may simply have no appropriate rights). I think that for this particular operation new SYSDBA attachment should be created which must create appropriate tables, role (something like PLG$PROFILE) if it does not exist yet and grant access on that tables to that role.

What about transaction in which profiler checks presence & access to that table - both approaches (use users transaction vs. start new one) have some advantages, i.e. I do not think it's critical API issue.

Reply:
> I agree that metadata creation must not run under random user because in that case we can't correctly control access to it (BTW, that user may simply have no appropriate rights). I think that for this particular operation new SYSDBA attachment should be created which must create appropriate tables, role (something like PLG$PROFILE) if it does not exist yet and grant access on that tables to that role.

Yes, I have similar idea.

> What about transaction in which profiler checks presence & access to that table - both approaches (use users transaction vs. start new one) have some advantages, i.e. I do not think it's critical API issue.

Expose user transaction to the unknown (untrusted ?) external plugin... hmm... may be not critical but definitely not safe, IMHO. More, now I see that allow plugin to use user attachment is also very far from safe.

Formally, profiler should be allowed to collect some stats and save it somewhere for further usage, correct ? 
I see no requirement here to provide profiler with a way to change user context or user data.
Ideally we should limit profiler access so that it is only allowed to create metadata and run DML on it.

Public API should not allow external plugin to do something that it should not do.

Reply:
Plugins are working inside of server and they receive master interface that allows them to connect any database including security one in embedded mode bypassing any security checks. There is no point to close the stable door.

Reply:
> BTW, I'm not sure profiler metadata should be owned by random user who happens to be first to use profiler.

I also do not like it.

> It is better to use separate account for it and make all necessary grants for public (or less wide) usage, IMHO.

Is it 100% safe to get database name with `rdb$get_context('SYSTEM', 'DB_NAME')` in server and reconnect to it with `SYSDBA` user and no password in the server?

Reply:
> Plugins are working inside of server and they receive master interface that allows them to connect any database including security one in embedded mode bypassing any security checks. There is no point to close the stable door.

Yes. Plugins must be trusted by server admin. They can also crash the server.

Reply:
> Plugins are working inside of server and they receive master interface that allows them to connect any database including security one in embedded mode bypassing any security checks. There is no point to close the stable door.

Engine could pass special impl of master interface that not allows to make connections at all.